### DISCUSSION TITLE
Nodemailer doesn't send emails after deployment

### DISCUSSION THREAD

**Discussion by jjjuk** [vercel/vercel#4387] [General] [answered]
I'm making reset password by email with graphql prisma. Nodemailer sendMail doesn't send email after first fire of mutation, but it sends email from first mutation after second firing. Basicaly sendMail only works if I run it twice.

Sometimes It throws error: 

`"Unexpected token A in JSON at position 0"`

Here's a piece of code inside mutation resolver:

```
const auth = {
          type: 'oauth2',
          user: [my email],
          clientId: [my client id]',
          clientSecret: [my client secret],
          refreshToken: '[my refresh token]
      };
      
const transporter = nodemailer.createTransport({         
          host: 'smtp.gmail.com',
          port: 465,
          secure: true, 
          auth
      });

```

```

const mail = {
 from: [my email],
 to: [email], 
 subject: 'Password Reset',
 html: [my html]
    };

```

      
    transporter.sendMail(mail, (err, info) => {
          if (err) {
              throw new Error(`${err}`)
          } else {
             return console.log(JSON.stringify(info));
          }
      });

**Comment by mcsdevv** [comment, marked as answer]
Hello @jjjuk, thanks for raising this. This behavior is not expected.

You can reference [this guide](https://vercel.com/guides/sending-emails-from-an-application-on-vercel) for our recommended approaches.

**Reply by Prioe** [reply]
Hi @mcsdevv, did Vercel implement this behaviour recently? I am currently using nodemailer with an smtp-transport to send mail from a nextjs api route. And my setup is consistently working.

My setup looks like this:

```js
const transporter = nodemailer.createTransport({
  host: 'smtp.sendgrid.net',
  port: 465,
  secure: true,
  auth: {
    user: 'apikey',
    pass: '[my-sendgrid-apikey]'
  }
})
```

Do I have to expect that this won't work in the future?

**Reply by paulogdm** [reply]
Hello @Prioe ,

Since Now 1.0, we block all SMTP traffic. But it seems you are using SendGrid, which is an "API-like email service". So the SMTP request is not started by your serverless function, but by SendGrid's servers. So your setup will always work without any issues.

Doing the code below, for example, would never work:
```PHP
<?php
$to      = 'nobody@example.com';
$subject = 'the subject';
$message = 'hello';
$headers = 'From: webmaster@example.com' . "\r\n" .
    'Reply-To: webmaster@example.com' . "\r\n" .
    'X-Mailer: PHP/' . phpversion();

mail($to, $subject, $message, $headers);
?> 
```

**Reply by jsbimra** [reply]
how to implement with smtp with own custom domains or what alternative without using any paid service to shoot mail after deploying from vercel do we need to buy vercel pro plan to work?  @paulogdm

**Reply by leerob** [reply]
You can read more about solutions here: https://vercel.com/docs/solutions/email

**Reply by Rakesh-99** [reply]
```
const transporter = nodemailer.createTransport({
                port: 465,
                host: "smtp.gmail.com",
                auth: {
                    user: USER,
                    pass: PASS,
                },
                secure: true,
            });

            await new Promise((resolve, reject) => {

                // verify connection configuration

                transporter.verify(function (error, success) {
                    if (error) {
                        console.log(error);
                        reject(error);
                    } else {
                        console.log("OTP has been sent to your Email");
                        resolve(success);
                    }
                });
            });

            // send mail with defined transport object

            const message = await transporter.sendMail({

                from: USER,
                to: req.body.email,
                subject: "Verify your Email ✔",
                html: `Your Email verification OTP IS : ${otp}`
            });

            // Send Mail :

            await new Promise((resolve, reject) => {

                transporter.sendMail(message, (err, info) => {
                    if (err) {
                        console.error(err);
                        reject(err);
                    } else {
                        console.log(info);
                        resolve(info);
                    }
                });
            });
        }
```

The Above code works fine in development(local), when deploy in vercel it doesn't send email. I am using nodemailer for this. Can you help me with that ? @leerob

**Comment by jsbimra** [comment]
Still sendgrid implementation not worked.

**Reply by leerob** [reply]
```js
import sendgrid from '@sendgrid/mail';

sendgrid.setApiKey(process.env.SENDGRID_API_KEY);

export default async (req, res) => {
  try {
    await sendgrid.send({
      to: 'your@email.com',
      from: 'your@email.com',
      subject: 'Serverless Functions',
      text: 'Hello, world!'
    });
  } catch (error) {
    return res.status(error.statusCode || 500).json({ error: error.message });
  }

  return res.status(200).json({ error: '' });
};
```

https://leerob.io/snippets/sendgrid

**Comment by jsbimra** [comment]
I have implemented this way too and still not receiving any emails and also
on vercel under function logs there is no error as now, also sendOrderEmail
api call it passed through with 200 but didn't receive any email.

how to debug further on vercel?

**Reply by Anoroc** [reply]
Assuming you've gone with Sendgrid, and without seeing your code, my suggestion would be to make sure you're function is running asynchronously. Without doing so will likely work locally, but not in production.

**Comment by jsbimra** [comment]
Yes it is defined asynchronously, on prod I have another file through email
is send successful but not through this file.

Error logged confirmed earlier in this email thread.

**Comment by jsbimra** [comment]
Okay good news is that things are working fine now without throwing any
errors on that file. you may close this ticket.

Thank you all for the support.

**Comment by jsbimra** [comment]
haha sorry, this is strange: I got the error back

[POST] /api/userLoginEmail19:00:40:15
2021-08-20T13:30:40.284Z 4f0bd9f0-6fac-4030-81f3-7f7733d4b8d0 ERROR
(node:8) Warning: To load an ES module, set "type": "module" in the
package.json or use the .mjs extension.(Use `node --trace-warnings ...` to
show where the warning was created)2021-08-20T13:30:40.288Z
4f0bd9f0-6fac-4030-81f3-7f7733d4b8d0 ERROR
/var/task/api/userLoginEmail.js:2import
Cookies from "cookies";^^^^^^SyntaxError: Cannot use import statement
outside a module at wrapSafe (internal/modules/cjs/loader.js:988:16) at
Module._compile (internal/modules/cjs/loader.js:1036:27) at
Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10) at
Module.load (internal/modules/cjs/loader.js:937:32) at
Function.Module._load (internal/modules/cjs/loader.js:778:12) at
ModuleWrap.<anonymous> (internal/modules/esm/translators.js:199:29) at
ModuleJob.run (internal/modules/esm/module_job.js:170:25) at async
Loader.import (internal/modules/esm/loader.js:178:24) at async getListener
(/var/task/___vc/__launcher.js:33:11)RequestId:
4f0bd9f0-6fac-4030-81f3-7f7733d4b8d0 Error: Runtime exited with error: exit
status 1Runtime.ExitError

**Comment by jsbimra** [comment]
Could this, would be the reason for the error?
{
  "functions": {
    "api/sendOrderEmail.js": {
      "includeFiles": "email-templates/**"
    },
    "api/userLoginEmail.js": {
      "includeFiles": "email-templates/**"
    }
  }
}

I guess here it could be through this. what is the syntax to add multiple
serverless definitions, I looked docs but just have single file example.

**Reply by jsbimra** [reply]
appreciate a quick interference to turn around this. Thanks!

**Comment by Olu-Tobi** [comment]
Check out this api, it worked for me https://rapidapi.com/tobidaramola77/api/mail-man/
